[2단계] 구현 완료했습니다. 리뷰 부탁드립니다

java-saeng opened the pull request:
# 미션 설명
안녕하세요!
이번 QnA 미션은 ATDD를 이용하여 질문 CURD를 구현하는 것입니다.

질문 생성
- 로그인 한 유저만 가능

질문 수정
- 로그인 한 유저가 질문 작성자일 경우 가능

질문 조회
- 모든 사용자가 가능

질문 삭제
- 로그인 한 유저가 질문 작성자일 경우 가능

# 과정
1. 하나의 기능에 대해 시나리오를 작성
2. 이를 토대로 acceptance test를 작성
3. controller unit test를 작성
4. controller 로직 작성
5. service unit test 작성
6. service 로직 작성
7. acceptance test 성공

# 정리

저번 리뷰 때 말씀해주셨던 JUnit4 -> JUnit5 업그레이드 했습니다.
버젼을 업그레이드하여 버젼 차이에 따라 코드를 변경하는 것은 정말 어려웠지만 좋은 경험이 됐다고 생각합니다.

저번 미션에 비해 원활하게 ATDD를 작성했다고 생각합니다
그러나 controller나 service에서 MockMvc, Mockito를 처음 사용해봐서 이 부분에 어려움을 느꼈습니다
이 부분을 따로 정리해서 공유하도록 하겠습니다!

# 궁금한 점
controller unit test에는 많은 방법이 있겠지만 제가 조사한 바로는 대표적으로 
1)MockMvc standaloneSetup, 2) MockMvc & @WebMvcTest 가 있다고 알고 있습니다.

현재 제가 사용한 방법은 1번인데, 이때는 @WebMvcTest가 필요없다고 생각합니다.


@WebMvcTest은 spring application context을 불러올텐데(실제 웹 서버 구동 X) 이것도 unit test라고 할 수 있나요??

이 부분도 unit test를 명확히 어떻게 정의할 수 없는 것과 비슷한 맥락일까요??

java-saeng committed:
feat : JUnit5 설정

java-saeng committed:
feat : 질문 생성 폼 ATDD

1. QuestionAcceptanceTest에서 해당 URI에 대한 요청하고 body를 출력하고 OK라는 응답을 얻는지 확인
2. QuestionController에서 대상 로직 구현

java-saeng committed:
test : 질문 생성 폼 요청 MockMvc 테스트

java-saeng committed:
feat : 질문 생성 ATDD

java-saeng committed:
feat : 질문 목록 조회 ATDD

java-saeng committed:
feat : 질문 단건 조회 ATDD

java-saeng committed:
feat : 질문 수정 폼 요청 ATDD

java-saeng committed:
feat : 질문 수정 실패 ATDD

java-saeng committed:
feat : 질문 수정 성공 ATDD

java-saeng committed:
feat : 질문 삭제 실패 ATDD

java-saeng committed:
feat : 질문 삭제 성공 ATDD

java-saeng committed:
fix : QnaServiceTest 테스트 오작동 해결 및 테스트의 DisplayName 수정

QnaServiceTest의 update_fail_not_match_user 테스트 실패 해결

java-saeng committed:
chore : DisplayName 변경

java-saeng committed:
feat : BeforeEach를 이용하여 코드 줄임

java-saeng committed:
refactor : createUser, createQuestion 변경 및 코드 줄이기

java-saeng committed:
refactor : createUser, createQuestion 변경 및 BeforeEach를 사용하여 코드 줄이기

java-saeng committed:
refactor : 코드 비교를 위한 수정

java-saeng committed:
style : code reformat

java-saeng commented on src/test/java/codesquad/web/QuestionControllerTest.java:
```diff
@@ -0,0 +1,208 @@
+        //given
+        //question, user 모두 id를 1로 하거나, 각각 1로 해도 오류가 뜸
+        Question question = createQuestion(0L);
+        User user = createUser(0L);
```
아이디를 1로 설정하지 않고 테스트를 한 이유는 실제 repository에 저장되는 것이 아니기 때문에 
question이 저장되고 나서 반환되는 아이디는 0이어서 0으로 설정했습니다.

실제 repository에 접근하여 테스트했다면 1을 반환했을 것이라고 생각합니다

hyukjin-lee replied:
서술된 내용의 의미를 정확하게 이해 못했습니다.
좀 더 모르는 사람에게 설명한다고 생각하고 부연설명해주시면 좋을 것 같아요

java-saeng replied:
해당 테스트는 controller layer의 unit test로 질문을 생성할 때, test-double을 통해 검증하기 때문에 실제 repository에 저장하지 않습니다. 

따라서 QuestionController에서 create 메서드를 이용하여 질문을 저장할 때는 반환되는 question의 id값이 0이 되기 때문에, 테스트가 정상적으로 작동하기 위해서는 아이디를 0L로 했습니다.

아래의 오류는 question과 user 모두 아이디를 1L로 설정했을 때 입니다
test-double에서 아이디가 1L인 question이 반환될 것이라 생각했지만 QuestionController의 create메서드를 통해 질문을 생성하면 0이 반환되는 것을 알 수 있습니다.

![image](https://user-images.githubusercontent.com/62413589/168511046-82879249-13b1-46d8-b0ab-aede1f75a8e7.png)

hyukjin-lee replied:
> 따라서 QuestionController에서 create 메서드를 이용하여 질문을 저장할 때는 반환되는 question의 id값이 0이 되기 때문에, 테스트가 정상적으로 작동하기 위해서는 아이디를 0L로 했습니다.

Service 의 create 메서드를 mocking 해주셨는데 본인이 케이스를 결정할 수 있는거 아닌가요~?
이해가 아직도 잘 되지 않아서 여쭙습니다..

java-saeng replied:
![image](https://user-images.githubusercontent.com/62413589/168949404-cbe4c4d8-8f65-4556-b738-8a17d15f63f5.png)
위 부분을 추가시켜줘서 아이디가 1L일 때 성공시켰습니다

해당 테스트는 "저장이 잘 되는가"가 아닌 "어떤 요청을 보낼 경우 홈으로 리다이렉트하는가?"에 대해 테스트하는 것이기 때문에 저희는 실제 question을 repository를 저장하는 것이 아니기 때문에 id를 param으로 보내줬어야했습니다!

java-saeng commented on src/test/java/codesquad/web/QuestionControllerTest.java:
```diff
@@ -0,0 +1,208 @@
+import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;
+
+@WebMvcTest(QuestionController.class)
```
standaloneSetup이 있으면 해당 어노테이션이 없어서 테스트가 실행될 것이라 생각합니다.

@WebMvcTest가 무엇인지, standaloneSetup이 무엇인지 따로 정리해서 공유하도록 하겠습니다.

java-saeng commented on src/test/java/codesquad/web/QuestionAcceptanceTest.java:
```diff
@@ -0,0 +1,194 @@
+        //when
+        ResponseEntity<String> response = basicAuthTemplate().postForEntity(savedQuestion.generateUrl(), htmlFormDataBuilder.build(), String.class);
+
+        log.info("response = {}", response);
```
http 스펙에서 put 요청을 할 때, 수정된 내용이 있으면 body로 반환한다고 알고 있습니다. 

response.getBody를 하면 null 값이 나오게 되는데, body로 반환하게 하는 것은 제가 따로 config를 작성해서 해야하나요??

hyukjin-lee replied:
> http 스펙에서 put 요청을 할 때, 수정된 내용이 있으면 body로 반환한다고 알고 있습니다.

그런가요? 혹시 관련 자료도 첨부해주시면 감사하겠습니다. 제가 알고 있던 내용이랑 다른 것 같아서요
https://developer.mozilla.org/ko/docs/Web/HTTP/Methods/PUT

java-saeng replied:
![image](https://user-images.githubusercontent.com/62413589/168503514-8b8385ab-b5e6-4fd2-9c08-e976d3e6650e.png)
위키 : https://ko.wikipedia.org/wiki/HTTP
stackoverflow : https://stackoverflow.com/questions/797834/should-a-restful-put-operation-return-something

저는 wiki를 보고 body가 있다고 생각했었습니다. stackoverflow를 보니 put은 response body가 필요하지 않다고 하네요

이 점이 좀 달라서 RFC 7231의 PUT을 읽어보았는데 response body가 명시적으로 필요하지 않다라는 말은 없었습니다.
다만, "When a PUT    representation is inconsistent with the target resource, the origin server SHOULD either make them consistent, by transforming the representation or changing the resource configuration, or respond with an appropriate error message containing sufficient information to explain why the representation is unsuitable." 이러한 구문이 있었는데, 제 생각에는 성공 시에는 status code만을 보내어 성공을 나타내고, 실패하면 왜 실패했는지 body에 내용을 담아 response를 하는 것으로 이해했습니다

java-saeng commented on src/test/java/codesquad/service/QnaServiceTest.java:
```diff
@@ -0,0 +1,183 @@
+        updatedQuestion.setContents("contents");
+
+        User user = createUser(2L);
+        user.setId(2L);
```
로컬에서는 해당 부분이 지워졌는데 무슨 이유인지는 모르겠지만 pr 날린 코드에는 안 지워져있습니다...

hyukjin-lee commented on src/main/java/codesquad/domain/Question.java:
```diff
@@ -77,6 +78,17 @@ public boolean isDeleted() {
         return deleted;
     }
 
+    public Question update(QuestionDto updatedQuestionDto) {
```
도메인 모델이 DTO 를 알게되면 어떤 단점이 생길까요?

java-saeng replied:
"도메인에서 DTO 로 변환 책임을 가지면 어떤 단점이 있을까요" 내용와 같이 Dto의 스펙이 바뀌면 그에 의존하는 엔티티의 코드가 바뀔 수 있다는 단점을 가지고 있다고 생각합니다

hyukjin-lee replied:
말씀해주신 코멘트에 덧붙여서 리뷰 드릴게요 ~

updatedQuestionDTO 는 클라이언트 쪽의 관심사를 담고 있는 데이터 구조체입니다. 
Question 은 백엔드 시스템의 도메인 모델인데, 
둘은 프로젝트 초기엔 유사한 형태를 띌 수는 있고 1:1 로 변환되거나 할 수 있지만
규모가 커질수록 서로 다른 형태의 모습을 띄게 될 가능성이 커지고 n:1 의 변환이 도메인에서 일어나게 될 가능성이 생깁니다.

화면의 요구사항은 매우 긴급하게 바뀌지만, 한번 정해진 도메인의 정책들은 쉽게 바뀌지 않습니다.
자주 변하는 것과 덜 자주 변하는 것을 분리하면, 중요한 것이 덜 중요한 변경사항에 영향 받는 것을 줄일 수 있습니다.
(세상을 잘 살아가는데에도 필요한 원리인 것 같습니다 ^^)

hyukjin-lee commented on src/main/java/codesquad/service/QnaService.java:
```diff
@@ -36,14 +39,33 @@ public Optional<Question> findById(long id) {
+        Question savedQuestion = questionRepository.findById(id)
+                                                   .orElseThrow(NoSuchElementException::new);
+
+        if(!savedQuestion.isOwner(loginUser) || loginUser == null){
```
loginUser == null 은 43 위로 올릴 수 있지 않을까요?

java-saeng replied:
말씀대로 loginUser == null을 DB 접근 전에 입력할 수 있습니다.

저는 아래 메서드인 deleteQuestion 과 예외 조건이 같은데 코드를 다르게 입력한 이유는 리뷰어님께 두 예외를 중 어느 코드가 더 깔끔한지 여쭤보고 싶었습니다.

deleteQuestion 메서드와 update 메서드의 차이는 loginUser가 null일 경우 데이터베이스에 접근 여부입니다.

loginUser == null 과 !savedQuestion.isOwner(loginUser) 둘 다 CannotDeleteException을 반환하는데, 저는 여기서 실패 조건을 하나로 합치면 코드 보기가 더 쉽지 않을까라는 생각을 했습니다.

하지만 이보다는 데이터베이스 접근이 무거운 작업이기 때문에 loginUser를 미리 검사하는 deleteQuestion의 코드가 더 알맞겠죠?? 
그리고 !savedQuestion.isOwner(loginUser)와 loginUser의 예외 원인은 서로 다르기 때문에 다르게 써주는게 나아보입니다

hyukjin-lee replied:
네 DB 접근을 줄일 수 있는게 최우선이 되어야겠죠. 코드 가독성을 위해 어느정도 성능은 감수할 수 있다고들하지만,
이런 중요한 부분은 DB 에도 영향을 주기 때문에 우선적으로 고려되어야할 것 같습니다

hyukjin-lee commented on src/main/java/codesquad/service/QnaService.java:
```diff
@@ -36,14 +39,33 @@ public Optional<Question> findById(long id) {
-    public Question update(User loginUser, long id, Question updatedQuestion) {
-        // TODO 수정 기능 구현
-        return null;
+    public Question update(User loginUser, long id, QuestionDto updatedQuestionDto) throws UnAuthenticationException {
```
throws 와 throw 의 차이점은 뭘까요~?

java-saeng replied:
이 부분은 정리해서 슬랙에 공유하도록 하겠습니다!

hyukjin-lee commented on src/main/java/codesquad/service/QnaService.java:
```diff
@@ -36,14 +39,33 @@ public Optional<Question> findById(long id) {
     public void deleteQuestion(User loginUser, long questionId) throws CannotDeleteException {
-        // TODO 삭제 기능 구현
+        if(loginUser == null){
+            throw new CannotDeleteException("삭제하지 못한다");
```
왜 삭제하지 못하는지 설명하는게 낫지 않을까요?

hyukjin-lee commented on src/main/java/codesquad/web/dto/QuestionDto.java:
```diff
@@ -0,0 +1,30 @@
+package codesquad.web.dto;
+
+import lombok.*;
```
wildcard 가 적용된 것 같습니다~

java-saeng replied:
"problem of using lombok wildcard"라는 검색어를 통해 구글링을 해봤지만 마땅한 답이 나오지 않아 질문드립니다!

wildcard를 얘기하신 이유가 lombok은 외부 라이브러리인데, 굳이 사용하지 않는 다른 lombok까지 import 해서 애플리케이션 실행 속도에 영향을 미치기 때문일까요~?

hyukjin-lee replied:
intellij wildcard 라고 검색하시면 원하는 답변들을 보실 수 있을 것 같아요~

java-saeng replied:
속도보다는 별개의 두 패키지를 동일한 이름의 클래스를 추가했을 경우에 어느 클래스를 참조할지 모르기 때문에 컴파일 시 문제가 생길 수도 있군요,, 처음 알았습니다 감사합니다!!

hyukjin-lee commented on src/main/java/codesquad/web/dto/QuestionDto.java:
```diff
@@ -0,0 +1,30 @@
+    private String contents;
+
+    @Override
+    public boolean equals(Object o) {
```
재정의 해준 이유가 있을까요?

java-saeng replied:
QuestionControllerTest의 update_success메서드애서 test-double을 통해 test를 진행하고 있습니다.

stubbing에서 QuestionDto의 모든 필드 값이 같지만, 서로 다른 객체이므로 mismatch 오류가 발생했습니다. 그래서 equals를 재정의해주었습니다.

아래는 equals를 재정의하지 않았을 때 생기는 오류입니다
![image](https://user-images.githubusercontent.com/62413589/168516581-1a3ac226-ad20-4581-96e2-8a3a649f9c4e.png)

hyukjin-lee replied:
테스트 통과를 위해 프로덕션 코드의 equals and hashcode 를 재정의하는게 맞을까요?

이 문제의 원인부터 천천히 생각해보고 어떤 방법을 취할 수 있는지 고민해보면 좋을 것 같습니다.

hyukjin-lee commented on src/test/java/codesquad/service/QnaServiceTest.java:
```diff
@@ -0,0 +1,183 @@
+    @Mock
+    private QuestionRepository questionRepository;
+
+    @InjectMocks
```
Mockito annotation 사용하지 않고 라이브러리 api 를 사용도 해보시면 좋을 것 같아요

java-saeng replied:
questionRepository만 mockito api 사용, QnaService mockito annotation 사용
![image](https://user-images.githubusercontent.com/62413589/168709443-fc1a1b10-fb63-48f3-bbba-d7d30c8b313c.png)

QnaService에 QuestionRepository 주입하려고 하는데, 필드 주입을 사용하더라구요 그래서 생성자를 만들어줘서 아래와 같이 작성하고 테스트를 작성했는데 잘 실행됐습니다
![image](https://user-images.githubusercontent.com/62413589/168710804-5d93d22e-565a-425c-9eae-ebcad524f919.png)

이래서 왜 필드 주입을 사용하게 되면 테스트 작성에 어려움이 생기는지 몸소 느낄 수 있었습니다!!

혹시 이 부분을 의도하시고 라이브러리 api를 사용해보라고 말씀하신걸까요~~???

hyukjin-lee replied:
이 방식이 좀 더 프레임워크의 어노테이션에 종속적이지 않고 일반적인 자바 코드로 테스트를 이해하기 더 편한 것 같아서요 ^^
저는 Mockito 어노테이션을 잘 사용하지 않습니다.

hyukjin-lee commented on src/test/java/codesquad/web/ApiUserAcceptanceTest.java:
```diff
@@ -11,7 +11,7 @@
 public class ApiUserAcceptanceTest extends AcceptanceTest {
 
     @Test
-    public void create() throws Exception {
```
junit 5 의 구성에 대해서 설명해주실 수 있을까요?

java-saeng replied:
junit5는 세 가지 하위 프로젝트인 JUnit Platform, JUnit Jupiter, Junit Vintage로 이뤄져있고, 이 아래 여러가지 모듈로 구성됩니다

JUnit Platform
- JUnit를 통해 테스트를 검색하고 실행할 수 있는 TestEngine API를 정의합니다. 모든 test engine을 통합해서 테스트를 돌리게 해주는 친구

JUnit Jupiter
- JUnit 5에서 테스트를 실행하기 위함

JUnit Vintage
- JUnit5 플랫폼에서 JUnit3, JUnit4도 지원해준다

hyukjin-lee commented on src/test/java/codesquad/web/QuestionAcceptanceTest.java:
```diff
@@ -0,0 +1,194 @@
+        htmlFormTestDataBuilder.addParameter("title", "title");
+        htmlFormTestDataBuilder.addParameter("contents", "aaaa");
+
+        htmlFormDataBuilder = htmlFormTestDataBuilder;
```
builder 를 builder 답게 필요한 곳에서 체이닝 방식으로 써보면 어떨까요~?

java-saeng replied:
오 감사합니다 생각하지도 못했습니다!!

hyukjin-lee commented on src/main/java/codesquad/domain/Question.java:
```diff
@@ -86,4 +98,8 @@ public String generateUrl() {
         return "Question [id=" + getId() + ", title=" + title + ", contents=" + contents + ", writer=" + writer + "]";
     }
+
+    public QuestionDto toDto() {
```
도메인에서 DTO 로 변환 책임을 가지면 어떤 단점이 있을까요

java-saeng replied:
제 생각에는 엔티티가 DTO에 의존하게 되면 DTO의 스펙이 바뀌게 되면 엔티티가 수정되야하기 때문에 좋지 않은 설계라고 생각합니다.
즉, DTO에 A라는 필드가 삭제되거나 새로운 B라는 필드가 추가되면은 엔티티에 있는 코드가 수정되야합니다. 즉, 외부 요인으로 인해 뷰가 바뀌고, 이로 인해 DTO가 바뀔 때마다 엔티티가 바뀌는 그런 상황이 생기기 때문에 아주 큰 단점을 가지게 됩니다

위 내용은 제 생각인데 혹시 관련 키워드를 알 수 있을까요,,? "entity has dependency dto" 요런 뉘앙스로 계속 검색했는데 명확한 답이 안나오는 것 같습니다

java-saeng replied:
이와 관련해서 의존성을 해결하기 위해 map struct, modelmapper가 있는 것으로 알고 있습니다!!

hyukjin-lee replied:
그럼 어디서 dto 변환을 해주는게 나을까요?
구글에 'dto transfrom layer' 라고 쳐봤는데 관련 게시물들이 나오긴 하네요

java-saeng replied:
결론부터 말씀드리자면 controller에서 변환하는 것이 나을 것 같습니다.

service layer에서는 해당 도메인에 대한 비즈니스 로직을 처리하는 layer라고 생각하는데, 변환까지 service가 하게 되면은 service가 너무 무거워질 것 같고, 모든 기능이 집중되있는 느낌이었습니다 

그래서 변환은 controller에서 하는게 좋다고 생각합니다

저는 map struct를 사용하여 dto와 entity 변환을 해주었습니다.
이와 관련해서 정리 해보고 다음 주 월요일에 한번 논의해보고 싶습니다

hyukjin-lee commented on src/test/java/support/test/AcceptanceTest.java:
```diff
@@ -2,14 +2,16 @@
 import org.springframework.test.context.junit4.SpringRunner;
 
-@RunWith(SpringRunner.class)
+@ExtendWith(SpringExtension.class)
```
@ExtendWith 가 @SpringBootTest 로 통합된게 버전 몇부턴지 기억이 잘 안나네요
한번 알아보시고 필요없는 버전이면 제거하셔도 될 것 같습니다.

java-saeng replied:
넵 현재 미션에서 사용한 spring boot 는 2.0.3으로 @SpringBootTest에 @ExtendWith이 포함되어있지 않더라구요!! 
첫번째 사진이 2.0.3의 @SpringBootTest, 두번째 사진이 2.6.7의 @SpringBootTest 입니다
![image](https://user-images.githubusercontent.com/62413589/168508284-603c97f8-cf3a-49be-a1dc-78d24d33e23f.png)
![image](https://user-images.githubusercontent.com/62413589/168508288-b14a79b2-56f1-4eb6-b57b-36605337e383.png)

java-saeng replied:
pending

hyukjin-lee reviewed (approved):
ATDD 로 짜려고 노력하신게 눈에 보이네요.
고생하셨습니다
리뷰 답변/반영하고 머지하시면 될 것 같습니다

java-saeng committed:
refactor : [리뷰 반영] builder를 builder답게 체이닝 방식 이용

java-saeng committed:
refactor : [리뷰 반영] 코드의 가독성보다는 DB 접근을 생각해보자

java-saeng committed:
refactor : [리뷰 반영] create 테스트에서 아이디를 1L로 설정해도 테스트 성공

java-saeng committed:
feat : map struct 기능 추가

java-saeng committed:
feat : map struct 기능 추가에 따른 코드 변경(toDto() 삭제)

Question
- DTO 의존 삭제

QnaService
- 생성자 주입으로 변경
- questionMapper 사용

java-saeng committed:
refactor : [리뷰 반영] intellij wildcard 수정

java-saeng committed:
feat : [리뷰 반영] DTO <=> Entity 변환은 controller에서 수행

java-saeng committed:
fix : post, put 요청 시 파라미터가 아닌 바디에 데이터를 전송

- 기존 코드에서는 파라미터를 통해 데이터를 전송
- post, put 요청 시 데이터는 바디에 실려야 한다
- form data는 key1=value1&key2=value2 형태로 바디에 데이터를 저장
- 이러한 형식을 지켜 mock 요청을 보냄
